torch/csrc/autograd/python_engine.cpp

```diff
@@ -31,13 +31,13 @@ static Engine& get_python_engine() {
 
 namespace torch { namespace autograd { namespace python {
 
-void PythonEngine::thread_init(int device) {
+void PythonEngine::thread_init(int device, const std::shared_ptr<ReadyQueue>& ready_queue) {
```

Comment:
While we're looking at memory lifetime issues, would you mind changing this from
  const std::shared_ptr<ReadyQueue> & 
to
  std::shared_ptr<ReadyQueue>

Same thing with reentrant_thread_init.

I think I can reason about the memory lifetime of the device ready queues (since we keep a vector of the shared_ptr around), though the other is harder to reason about-  we might be ok, might not.

This ends up only adding on extra refcount inc/dev per spawned engine thread, fairly cheap insurance against memory lifetime issues here.